Converter for functor system

GiudGiud opened the pull request:
anticipated the disappearance of declareFunctorProperty and just called it FunctorADConverter instead of FunctorMaterialConverter

GiudGiud committed:
Prevent a common failure of GenericFunctorMaterial

GiudGiud committed:
Rename functor by block material
Refs #16809

GiudGiud committed:
Re-add piecewise constant by block regular material, as some users need it

GiudGiud committed:
Add a functor material property converter between AD and regular
refs #19420

GiudGiud committed:
Rename MaterialConverter to MaterialADConverter

GiudGiud committed:
Remove extra multipliciations by mu rampdown
lang syntax

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/903974/) on 313e242 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19512/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19512/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 97d5b7b17e145d7aa912802bff51f34763dda2df`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/904008/) on b0f183c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19512/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/904035/) on b0f183c wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>97d5b7</th>
      <th colspan=4>#19512 b0f183</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.07%</td>
      <td>82.06%</td>
      <td>-0.01%</td>
      <td>54.25%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72516</td>
      <td>72581</td>
      <td>+65</td>
      <td>83</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15846</td>
      <td>15870</td>
      <td>+24</td>
      <td>70</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19512/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19512/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>97d5b7</th>
      <th colspan=4>#19512 b0f183</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.16%</td>
      <td>83.16%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7601</td>
      <td>7599</td>
      <td>-2</td>
      <td>12</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1539</td>
      <td>1539</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19512/coverage_diff/navier_stokes.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19512/coverage/navier_stokes)

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>97d5b7</th>
      <th colspan=4>#19512 b0f183</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.05%</td>
      <td>89.05%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>24854</td>
      <td>24854</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3055</td>
      <td>3055</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19512/coverage_diff/tensor_mechanics.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19512/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19512/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 54.25% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud committed:
Add more coverage on regular matprop constant per block material

GiudGiud committed:
Add documentation for functor conversion material and piecewise constant material, refs #19420
Add exception checks test to functor conversion material

GiudGiud committed:
Add ghosting to pass the boundary test for block material
Ideally, we would use a postprocessor that currently does not have an issue with insufficient
ghosting when used on its own, see #19534. But I dont think there is one that will trigger
the ghosting we want. I could admittedly make a new one, which would be a rework of SideAverageMaterialProperty,
but it could make that one confusing.

Use MatpropName and FunctorName instead of strings for parameters for the AD converters
Refs #19420

Use interface diffusive flux instead of side to use both sides

regold

dschwen commented on framework/src/materials/FunctorADConverter.C:
```diff
@@ -0,0 +1,106 @@
+FunctorADConverterTempl<T>::FunctorADConverterTempl(const InputParameters & parameters)
+  : FunctorMaterial(parameters)
+{
+  auto reg_props_in = getParam<std::vector<std::string>>("reg_props_in");
```
It'd be more elegant to use `getParam<std::string, std::string>("reg_props_in", "ad_props_out");` to have MOOSE build a vector of pairs

GiudGiud replied:
could do a map too, but all the Generic Materials are not built this way

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/907049/) on bcd048a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19512/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19512/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 810ea0ea6a06362314d22cd0699024f9b60a5235`

GiudGiud committed:
Remove all remaining occurences of NodalMaxValue, refs #18938
Sloppy omission, need to fix vector type

GiudGiud commented:
Regular ADConverter, which I've only renamed, is covered in tensor_mechanics. I think there's a case for merging with poor coverage haha

GiudGiud committed:
Add tests within framework for MaterialADConverter

Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented:
Ok added a within framework test for materialAdConverter. 
Also should get better coverage for the piecewiseConstantByBlockMaterial now

lindsayad commented on framework/doc/content/source/materials/FunctorADConverter.md:
```diff
@@ -0,0 +1,28 @@
+
+## Example input syntax
+
+An example of some gymnastic with functor conversions is shown in this example. The reader should note
```
```suggestion
An example of some gymnastics with functor conversions is shown in this example. The reader should note
```

lindsayad commented on framework/src/materials/FunctorADConverter.C:
```diff
@@ -0,0 +1,107 @@
+{
+  InputParameters params = FunctorMaterial::validParams();
+  params.addClassDescription("Converts regular functors to AD functors and "
+                             " AD functors to regular functors");
```
seems like an unnecessary space here?

GiudGiud replied:
```suggestion
                             "AD functors to regular functors");
```

lindsayad commented on framework/src/materials/FunctorADConverter.C:
```diff
@@ -0,0 +1,107 @@
+  // Define the AD functors
+  for (const auto i : make_range(reg_props_in.size()))
+  {
+    const auto & reg_functor = &getFunctor<T>(reg_props_in[i]);
+    const auto & ad_prop =
+        &declareFunctorProperty<typename Moose::ADType<T>::type>(ad_props_out[i]);
+    ad_prop->setFunctor(_mesh,
+                        blockIDs(),
+                        [reg_functor](const auto & r, const auto & t) ->
+                        typename Moose::ADType<T>::type { return (*reg_functor)(r, t); });
```
```suggestion
    const auto & reg_functor = getFunctor<T>(reg_props_in[i]);
    auto & ad_prop =
        declareFunctorProperty<typename Moose::ADType<T>::type>(ad_props_out[i]);
    ad_prop.setFunctor(_mesh,
                        blockIDs(),
                        [&reg_functor](const auto & r, const auto & t) ->
                        typename Moose::ADType<T>::type { return reg_functor(r, t); });
```

lindsayad commented on framework/src/materials/FunctorADConverter.C:
```diff
@@ -0,0 +1,107 @@
+  // Define the regular functors
+  for (const auto i : make_range(ad_props_in.size()))
+  {
+    const auto & ad_functor = &getFunctor<typename Moose::ADType<T>::type>(ad_props_in[i]);
+    const auto & reg_prop = &declareFunctorProperty<T>(reg_props_out[i]);
+    reg_prop->setFunctor(_mesh, blockIDs(), [ad_functor](const auto & r, const auto & t) -> T {
+      return MetaPhysicL::raw_value((*ad_functor)(r, t));
```
```suggestion
    const auto & ad_functor = getFunctor<typename Moose::ADType<T>::type>(ad_props_in[i]);
    auto & reg_prop = declareFunctorProperty<T>(reg_props_out[i]);
    reg_prop.setFunctor(_mesh, blockIDs(), [&ad_functor](const auto & r, const auto & t) -> T {
      return MetaPhysicL::raw_value(ad_functor(r, t));
```

lindsayad reviewed (changes requested):
Why do we even want `PiecewiseConstantByBlockMaterial` any more? It was previously converted to be a functor material so any users of that material should already be accepting functors instead of regular material properties, so it seems like anyone previously using it should just use `PicewiseByBlockFunctorMaterial` instead and that we should consequently have no current users of a regular material version of it.

GiudGiud commented:
We have a potential user in a discussions thread
In general it s a good idea to have this material, it s a shorter syntax than the existing options
https://github.com/idaholab/moose/discussions/19470

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/909805/) on d8b0429 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19512/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19512/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e5db31b8c2769f31148a2c4c4449b96404e93c38`
